File: 05_Graph_Search/Level3/1743.cpp
// Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
// 음식물 피하기
// https://www.acmicpc.net/problem/1743
// 힌트
// 1. 모든 위치를 탐색하다가 음식물이 발견되면 
//    해당 위치에서 상하좌우 위치로 DFS를 이용하여
//    음식물의 갯수를 모두 더해준다. 한번 카운팅했다면 0으로 만들어주어 다시 세지 않도록 한다.
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;
int M, N, K;
int answer = 0;
vector<vector<int> > table;

int dfs(int i, int j, int cnt)
{
    table[i][j] = 0;
    if (i - 1 > 0 && table[i - 1][j] == 1)
        cnt += dfs(i - 1, j, 1);
    if (j - 1 > 0 && table[i][j - 1] == 1)
        cnt += dfs(i, j - 1, 1);
    if (i + 1 <= N && table[i + 1][j] == 1)
        cnt += dfs(i + 1, j, 1);
    if (j + 1 <= M && table[i][j + 1] == 1)
        cnt += dfs(i, j + 1, 1);
    return cnt;
}
int main()
{
    cin >> N >> M >> K;
    answer = 0;
    table = vector<vector<int> >(N + 1, vector<int>(M + 1, 0));
    for (int i = 0; i < K; i++)
    {
        int a, b;
        cin >> a >> b;
        table[a][b] = 1;
    }
    for (int i = 1; i <= N; i++)
    {
        for (int j = 1; j <= M; j++)
        {
            if (table[i][j] == 1)
            {
                answer = max(answer, dfs(i, j, 1));
            }
        }
    }
    cout << answer << endl;
    return 0;
}

File: 05_Graph_Search/Level3/1743.py
# Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
# 연결 요소의 개수
# https://www.acmicpc.net/problem/11724
# 힌트
# 1. DFS를 활용하여 1~N까지의 연결 요소를 탐색하며, 염결된 정점이 있으면 방문하도록 하자.
# 2. 이때 한번 방문한 정점을 반복해서 방문하지 않도록 check 배열을 이용해 준다.
import sys
sys.setrecursionlimit(1000000)


def solve():
    ans = 0

    for i in range(1, N + 1):
        for j in range(1, M + 1):
            if table[i][j] == 1:
                ans = max(ans, dfs(i, j, 1))

    return ans


def dfs(i, j, cnt):
    table[i][j] = 0
    if i - 1 > 0 and table[i - 1][j] == 1:
        cnt += dfs(i - 1, j, 1)
    if j - 1 > 0 and table[i][j - 1] == 1:
        cnt += dfs(i, j - 1, 1)
    if i + 1 <= N and table[i + 1][j] == 1:
        cnt += dfs(i + 1, j, 1)
    if j + 1 <= M and table[i][j + 1] == 1:
        cnt += dfs(i, j + 1, 1)

    return cnt


if __name__ == "__main__":
    N, M, K = map(int, sys.stdin.readline().split())

    table = [[0] * (M + 1) for _ in range(N + 1)]
    check = [[False] * (M + 1) for _ in range(N + 1)]

    for i in range(K):
        a, b = map(int, sys.stdin.readline().split())
        table[a][b] = 1

    answer = solve()
    print(answer)